### ISSUE TITLE
Proposal: Add new property for getting process path

### ISSUE THREAD

**Issue by Giorgi** [dotnet/corefx#1081]
As you probably know you can get process path using `someProcess.MainModule.FileName` snippet. However it suffers from a couple of issue:
* Accessing MainModule of 64 bit process from a 32 bit one throws an exception
* [You can't access MainModule of elevated process from a non elevated one](http://stackoverflow.com/a/8601944/239438)

I suggest adding a new property call `Path` which returns the path to executable of the Process. To overcome the issues that I described we can use `QueryFullProcessImageName` together with `PROCESS_QUERY_LIMITED_INFORMATION` access right.

I am happy to send the Pull Request if you decide to go forward with this change.

**Comment by ellismg** [comment]
If we do this, on Linux, we could use /proc/<pid>/exe to figure out where the on disk image is.  There are some caveats listed in the man pages, but in general I think it would be OK.  Not sure how (or if) we can do this on OS X.

**Comment by mikedn** [comment]
Possibly related discussion/work in progress: https://github.com/dotnet/roslyn/pull/884#discussion-diff-25801767

**Comment by stephentoub** [comment]
> on Linux, we could use /proc//exe to figure out where the on disk image is.

Yeah, we already use this in our Process implementation on Linux. In order to do path resolution for the file name passed to Start, we use /proc/self/exe to get the path of the exe and look on its directory.

**Comment by Giorgi** [comment]
This has two ways for getting process path on Mac: [a](https://stackoverflow.com/questions/799679/programatically-retrieving-the-absolute-path-of-an-os-x-command-line-app/)

**Comment by xanather** [comment]
This is needed! Environment.ExecutingPath? Something like that. Ive always had to play with reflection, AppDomains (doesn't exist in .Net Core), or use System.Windows.Forms.Application.StartupPath (doesn't exist in .Net Core) to get such a simple thing.

**Comment by mikedn** [comment]
@xanather This issue is specific to the `System.Diagnostics.Process` class, you seem to be asking for something different.

>AppDomains (doesn't exist in .Net Core), 

There is [AppContext]( https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/AppContext/AppContext.cs#L22).

**Comment by terrajobst** [comment]
If `Process.MainModule` will end up throwing on certain platforms, we should also expose a query API, like `Process.CanQueryModuleInformation` or something along those lines.

**Comment by Giorgi** [comment]
It already throws exception if the target process is elevated and the caller isn't

**Comment by Priya91** [comment]
@giorgi Can you provide an api proposal, a good example [here](https://github.com/dotnet/corefx/issues/271)

**Comment by Giorgi** [comment]
The api is very simple:

```csharp    
 string Path { get; }
```

**Comment by Priya91** [comment]
@Giorgi The api proposal needs to be in the form of a speclet as mentioned in the link. You can edit your first comment on this issue, with the api speclet, mainly outlining on the usage scenarios to justify need of the API and the API structure like below. This will help during api review meeting to get to the necessary information quickly rather that perusing the whole issue.

Note: Path doesn't seem very obvious that it's indicating the main module file path, so renamed API below. Open to other suggestions.

```
namespace System.Diagnostics
{
     public class  Process : Component
     {
           public string MainModuleFilePath { get; }    
     }
}
```

**Comment by Giorgi** [comment]
@Priya91  Post updated.

**Comment by Priya91** [comment]
@Giorgi The proposal looks good, you have mentioned that it's possible to implement on windows, can you also update the post on similar investigations on Linux and OSX. .NETCore supports Linux and OSX, so we want to have data on whether this API returns meaningful value on other platforms as well.

**Comment by Giorgi** [comment]
@Priya91 Done!

**Comment by danmosemsft** [comment]
@Priya91 I marked as ready for review per your note above, feel free to flip back if there is still more feedback.

**Comment by terrajobst** [comment]
We just reviewed this today:

* I understand that accessing the main module currently throws. Any reason we cannot make this API work and have other properties on `ProcessModule` throw if the bitness differs or the process is elevated?
* The concern being, that adding multiple properties to access the same information might make the API more confusing.

**Comment by mellinoe** [comment]
> I understand that accessing the main module currently throws.

I don't think that's true, barring the caveats in the original post (e.g. 64->32-bit or non-elevated->elevated processes on Windows). We just recently started using it in [the corefx tests here](https://github.com/dotnet/corefx/blob/5bd6dea77374feccbc859d4a771f9fff7e973d86/src/Common/tests/System/Diagnostics/RemoteExecutorTestBase.cs#L21), and I do believe that code is being run on all of our platforms.

**Comment by Priya91** [comment]
> Any reason we cannot make this API work and have other properties on ProcessModule throw if the bitness differs or the process is elevated?

We can defer throwing exceptions in the implementation to when the properties are accessed, and use a different win32 API for ProcessModule.FileName, but that will cause a behavior change from desktop. And there could be code, for, if MainModule is null assume querying process is elevated so no access, or if MainModule throws, assume different bitness, hence do something else, etc.

**Comment by stephentoub** [comment]
From an implementation standpoint, in what situations is it possible to provide MainModuleFilePath and not MainModule.FilePath?  On macOS we simply construct a single ProcessModule from getting the file path (https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.OSX.cs#L113-L123), so if we can get the latter we can get the former.  Similarly for Linux, we get the module information by parsing a virtual file from procfs (https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Linux.cs#L51), so if we can get the file path, we can get the rest of the data.

**Comment by Giorgi** [comment]
@stephentoub I haven't tried it on other platforms but I do have encountered this issue on Windows. Here is this bug on Microsoft connect with a small repro and workaround: [https://connect.microsoft.com/VisualStudio/feedback/details/595717/add-path-property-to-process-class-so-that-it-works-with-elavated-processes-too](url)

**Comment by wtgodbe** [comment]
@Giorgi @Priya91 @stephentoub do we still plan on taking this change/was a PR ever put up for this issue? Would like to get closure one way or another.

**Comment by stephentoub** [comment]
The API is still marked "api-needs-work", which suggests additional follow-up is necessary before it can be re-reviewed.  And the milestone is set as future, which suggests there's no rush here.  What kind of closure are you looking for, and why?

**Comment by Giorgi** [comment]
@wtgodbe I haven't sent a PR because the suggested api was [rejected](https://github.com/dotnet/corefx/issues/1081#issuecomment-273277103) but I can't think of a better api for getting process path.

I think it would be very intuitive to have a `Path` property for accessing process path instead of using `MainModule.FileName` and it would also work in cases when accessing `MainModule` throws an exception.

**Comment by wtgodbe** [comment]
I will leave this open then, if you or someone else can come up with a proposal that addresses @terrajobst concerns then feel free to do so.

**Comment by Giorgi** [comment]
@wtgodbe As @Priya91 has said in her comment addressing @terrajobst concerns isn't possible without introducing breaking change so I don't think there is any other way.

**Comment by Giorgi** [comment]
@terrajobst @mellinoe @stephentoub What if we add an extension method in Windows Compatibility Pack ?
